sb3_contrib/common/distributions.py

```diff
@@ -0,0 +1,146 @@
+
+            logits = th.where(self.masks, self.logits, HUGE_NEG)
+        else:
+            self.masks = None
+            logits = self.logits
+
+        # Reinitialize updated logits
+        super().__init__(logits=logits)
+
+    def entropy(self):
+        if self.masks is None:
+            return super().entropy()
+
+        # Highly negative logits don't result in 0 probs, so we must replace
+        # with 0s to ensure 0 contribution to the distribution's entropy, since
+        # masked actions possess no uncertainty.
+        device = self.logits.device
+        p_log_p = self.logits * self.probs
+        p_log_p = th.where(self.masks, p_log_p, th.tensor(0.0, device=device))
+        return -p_log_p.sum(-1)
+
+
+class CategoricalDistribution(Distribution):
+    """
+    Categorical distribution for discrete actions.
+
+    :param action_dim: Number of discrete actions
+    """
+
+    def __init__(self, action_dim: int):
+        super().__init__()
+        self.distribution: MaskableCategorical = None
+        self.action_dim = action_dim
+
+    def proba_distribution_net(self, latent_dim: int) -> nn.Module:
+        """
+        Create the layer that represents the distribution:
+        it will be the logits of the Categorical distribution.
+        You can then get probabilities using a softmax.
+
+        :param latent_dim: Dimension of the last layer
+            of the policy network (before the action layer)
+        :return:
+        """
+        action_logits = nn.Linear(latent_dim, self.action_dim)
+        return action_logits
+
+    def proba_distribution(self, action_logits: th.Tensor) -> "CategoricalDistribution":
+        self.distribution = MaskableCategorical(logits=action_logits)
+        return self
+
+    def log_prob(self, actions: th.Tensor) -> th.Tensor:
+        assert self.distribution is not None, "Must set distribution parameters"
+        return self.distribution.log_prob(actions)
+
+    def entropy(self) -> th.Tensor:
+        assert self.distribution is not None, "Must set distribution parameters"
+        return self.distribution.entropy()
+
+    def sample(self) -> th.Tensor:
+        assert self.distribution is not None, "Must set distribution parameters"
+        return self.distribution.sample()
+
+    def mode(self) -> th.Tensor:
+        assert self.distribution is not None, "Must set distribution parameters"
+        return th.argmax(self.distribution.probs, dim=1)
+
+    def actions_from_params(self, action_logits: th.Tensor, deterministic: bool = False) -> th.Tensor:
+        # Update the proba distribution
+        self.proba_distribution(action_logits)
+        return self.get_actions(deterministic=deterministic)
+
+    def log_prob_from_params(self, action_logits: th.Tensor) -> Tuple[th.Tensor, th.Tensor]:
+        actions = self.actions_from_params(action_logits)
+        log_prob = self.log_prob(actions)
+        return actions, log_prob
+
+
+def make_proba_distribution(
```

Comment:
this should not be needed as `PPO_MASK` will only use `MaskedCategorical` distribution, no?

Reply:
Yes, but users could provide environments with any space they'd like. I wrote the algorithm as though masking was an optional opt-in feature on top of the existing `PPO`, instead of raising an exception if the user used the algorithm in a way that wouldn't result in masking.

I guess that's overly permissive, it's probably clearer to just raise the exception. If the user is happy to proceed without masking, they should just use SB3's `PPO`. If this ever became a native SB3 feature, it would be a different story.

That said, I'd like to add other distributions besides `Categorical`, at least `MultiCategorical` and `Bernoulli`. So this function would probably be useful in the future.